File: Day3/Endianess.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdint.h>

#define BYTE_ORDER_MARK 0xFFFE

int main() {
    union {
        uint16_t bom;
        uint8_t endianDetect[2];
    } cauldron;
    cauldron.bom = BYTE_ORDER_MARK;
    switch (cauldron.endianDetect[0]) {
    case 0xFF: puts("Big Endian"); break;
    case 0xFE: puts("Little Endian"); break;
    default: assert(false);
    }
}

File: Day03/Endianess.java
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class Endianess {
    private static final short BYTE_ORDER_MARK = (short) 0xFFFE;

    public static void main(final String... args) {
        var buffer = ByteBuffer.allocate(2);
        buffer.order(ByteOrder.nativeOrder());
        buffer.putShort(BYTE_ORDER_MARK);
        buffer.rewind();
        switch (buffer.get() & 0xFF) {
        case 0xFF: System.out.println("Big Endian"); break;
        case 0xFE: System.out.println("Little Endian"); break;
        default: assert false;
        }

        switch (ByteOrder.nativeOrder()) {
        case ByteOrder.BIG_ENDIAN: System.out.println("Big Endian"); break;
        case ByteOrder.LITTLE_ENDIAN: System.out.println("Little Endian"); break;
        default: assert false;
        }
    }
}
